File: 99.恢复二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=99 lang=cpp
 *
 * [99] 恢复二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


// #include <iostream>
// using namespace std;
// #include <vector>
// #include <algorithm>
// #include <map>

// struct TreeNode
// {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };


class Solution
{
public:

    vector<int> result;
    int count = 0;

    void inorderTrack(TreeNode *root)
    {
        if(root == NULL)
        {
            return ;
        }

        inorderTrack(root->left);
        result.push_back(root->val);
        inorderTrack(root->right);
    }

    void changeTrack(TreeNode * & root, int x, int y)//可能出错
    {
        if(root == NULL)
        {
            return ;
        }

        changeTrack(root->left, x, y);
        if(count == x)
        {
            root->val = result[x];
        }

        if(count == y)
        {
            root->val = result[y];
        }
        count++;
        changeTrack(root->right, x, y);
    }

    void recoverTree(TreeNode *root)
    {
        inorderTrack(root);
        vector<int> temp = result;
        sort(result.begin(), result.end());

        //map<int, int> now;
        vector<int> now;

        for(int i = 0; i < temp.size(); i++)
        {
            if(temp[i] != result[i])
            {
                //now.insert(pair<int, int>(i, 1));
                now.push_back(i);
            }
        }

        changeTrack(root, now[0], now[1]);
    }
};
// @lc code=end

File: 99.恢复二叉搜索树.py
#
# @lc app=leetcode.cn id=99 lang=python3
#
# [99] 恢复二叉搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from typing import Optional
from typing import List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    def __init__(self) -> None:
        self.result = []
        self.count = 0

    def inorderTrack(self, root:Optional[TreeNode]) ->None:
        if root is None :
            return
        
        self.inorderTrack(root.left)
        self.result.append(root.val)
        self.inorderTrack(root.right)

    def changeTrack(self, root: Optional[TreeNode], x:int, y:int) -> None :
        if root is None :
            return
        
        self.changeTrack(root.left, x, y)
        if self.count == x :
            root.val = self.result[x]

        if self.count == y :
            root.val = self.result[y]

        self.count += 1
        self.changeTrack(root.right, x, y)

    def recoverTree(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """

        self.inorderTrack(root)
        temp = self.result[:]
        self.result.sort()
        now = []

        for i in range(len(temp)):
            if temp[i] is not self.result[i] :
                now.append(i)
            
        self.changeTrack(root, now[0], now[1])
# @lc code=end

